File: Frontend/src/app/page/PaginatedResult.ts
import { count, Observable, skip, take, toArray } from "rxjs";

export class PaginatedResult<T> {
    pageIndex: number;
    totalPages: number;
    items: T[] ;
  
    constructor(items: T[], count: number, pageIndex: number, pageSize: number) {
      this.pageIndex = pageIndex;
      this.totalPages = Math.ceil(count / pageSize);
      this.items = items;
    }
  
    get hasPreviousPage(): boolean {
      return this.pageIndex > 1;
    }
  
    get hasNextPage(): boolean {
      return this.pageIndex < this.totalPages;
    }
  
    static async create<T>(
        query: Observable<T[]>,
        pageIndex: number,
        pageSize: number
      ): Promise<PaginatedResult<T>> {
        const [items, totalCount] = await Promise.all([
          query.pipe(
            skip((pageIndex - 1) * pageSize),
            take(pageSize),
            toArray()
          ).toPromise(),
          query.pipe(
            count()
          ).toPromise()
        ]) as [T[], number];
      
        return new PaginatedResult<T>(items, totalCount, pageIndex, pageSize);
      }

      
  }

File: Backend/WebSellingPhone.Bussiness/ViewModel/PaginatedResult.cs
using Microsoft.EntityFrameworkCore;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace WebSellingPhone.Bussiness.ViewModel
{
    public class PaginatedResult<T> where T : class
    {
        public int PageIndex { get; private set; }
        public int TotalPages { get; private set; }
        public T[] Items { get; set; }

        public PaginatedResult(List<T> items, int count, int pageIndex, int pageSize)
        {
            PageIndex = pageIndex;
            TotalPages =(int) Math.Ceiling(count * 1.0 / pageSize );

            Items = items.ToArray();
            Console.WriteLine(TotalPages.ToString());
        }

        public bool HasPreviousPage => PageIndex > 1;

        public bool HasNextPage => PageIndex < TotalPages;

        public static async Task<PaginatedResult<T>> CreateAsync(IQueryable<T> query, int pageIndex, int pageSize)
        {
            var count = await query.CountAsync();
            Console.WriteLine(count.ToString());
            var items = await query.Skip((pageIndex - 1) * pageSize).Take(pageSize).ToListAsync();
            return new PaginatedResult<T>(items, count, pageIndex, pageSize);
        }
    }
}
